# Hacker News Analysis

Hacker news is website, popular in technology and startup cicle. Hacker News gets hundreds of listings and thousands of visitors voting for the listings. People interested in posts/listings might raise the concerns and also submit some interesting project, product to a community. 

This dataset has 7 columns id: unique identifier for the posts at Hacker News, The title of the post, Url, The number of points the post acquired, number of comments on the post and the author who submitted the post as well as the date and time of the creation of post



In [3]:
import csv

with open('hacker_news.csv') as opened_file:
    read_file = csv.reader(opened_file)
    hn = list(read_file)
    
hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

In [4]:
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [5]:
ask_posts = []
show_posts = []
other_posts = []
for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(title)
    elif title.lower().startswith('show hn'):
        show_posts.append(title)
    else:
        other_posts.append(title)
    
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))
    
        

1744
1162
17194


### Calculating average number of comments 

In [6]:
total_ask_comments = 0 
total_show_comments = 0
other_posts = 0

for post in hn:
    title = post[1]  # Fix variable reference
    num_comments = post[4]
    num_comments = int(num_comments) 

    if title.lower().startswith('ask hn'):
        total_ask_comments += num_comments
    elif title.lower().startswith('show hn'):
        total_show_comments += num_comments
    else:
        other_posts += num_comments

if len(ask_posts) > 0:
    avg_ask_comments = total_ask_comments / len(ask_posts)
else:
    avg_ask_comments = 0

if len(show_posts) > 0:
    avg_show_comments = total_show_comments / len(show_posts)
else:
    avg_show_comments = 0

print(avg_ask_comments)
print(avg_show_comments)     

14.038417431192661
10.31669535283993


It was observed that the ask posts recieve more comments compared to show posts. 

#### Number of posts and comments by hour 

As observed ask posts receive more comments than show posts, I will focus onn ask posts for remaining analysis. Here I will determine if ask posts created at a certain time are more likely to attract comments by performing following steps 
1. Calculate the number of ask posts created in each hour of the day, along with the number of comments received.
1. Calculating the average number of comments ask posts receive by hour created.


In [7]:
import datetime as dt

result_list = []

for post in hn:
    title = post[1]
    date_time = post[6]  # Fix variable reference
    comment_post = post[4]
    comment_post = int(comment_post)
    if title.lower().startswith('ask hn'):
        result_list.append((date_time, comment_post))
    
comments_by_hour = {}
counts_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for each_row in result_list:
    date = each_row[0]
    comment = each_row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in counts_by_hour:
        comments_by_hour[time] += comment
        counts_by_hour[time] += 1
    else:
        comments_by_hour[time] = comment
        counts_by_hour[time] = 1

comments_by_hour 

{'09': 251,
 '13': 1253,
 '10': 793,
 '14': 1416,
 '16': 1814,
 '23': 543,
 '12': 687,
 '17': 1146,
 '15': 4477,
 '21': 1745,
 '20': 1722,
 '02': 1381,
 '18': 1439,
 '03': 421,
 '05': 464,
 '19': 1188,
 '01': 683,
 '22': 479,
 '08': 492,
 '04': 337,
 '00': 447,
 '06': 397,
 '07': 267,
 '11': 641}

### Calculating average number of comments for posts created during each hour of the day

In [11]:
avg_by_hour = []

for key in comments_by_hour:
    avg_by_hour.append([key, (comments_by_hour[key])/(counts_by_hour[key])]) 
    
avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

### Sorting and Printing values from a list of lists

In [15]:
swap_avg_by_hour = []

for row in avg_by_hour: 
    swap_avg_by_hour.append([row[1], row[0]])
    
# print(swap_avg_by_hour)
sorted_swap = sorted(swap_avg_by_hour, reverse=True)

sorted_swap 

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [17]:
print("Top 5 Hours for Ask posts Comments")
for avg,hr in sorted_swap:
    print(
        f"{dt.datetime.strptime(hr, '%H').strftime('%H:%M')}:{avg:.2f} average comments per post")
    
          

Top 5 Hours for Ask posts Comments
15:00:38.59 average comments per post
02:00:23.81 average comments per post
20:00:21.52 average comments per post
16:00:16.80 average comments per post
21:00:16.01 average comments per post
13:00:14.74 average comments per post
10:00:13.44 average comments per post
14:00:13.23 average comments per post
18:00:13.20 average comments per post
17:00:11.46 average comments per post
01:00:11.38 average comments per post
11:00:11.05 average comments per post
19:00:10.80 average comments per post
08:00:10.25 average comments per post
05:00:10.09 average comments per post
12:00:9.41 average comments per post
06:00:9.02 average comments per post
00:00:8.13 average comments per post
23:00:7.99 average comments per post
07:00:7.85 average comments per post
03:00:7.80 average comments per post
04:00:7.17 average comments per post
22:00:6.75 average comments per post
09:00:5.58 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set documentation, the timezone used is Eastern Time in the US. So, it can also written 15:00 as 3:00 pm est.

## Conclusion 

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on the analysis, to maximize the amount of comments a post receives, I'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.